### PR TITLE
[TypeDeclaration] Handle union A extends B on ReturnUnionTypeRector

### PR DESCRIPTION
**Pull request by samsonasik**
Cherry picked from PR:

- https://github.com/rectorphp/rector-src/pull/5148

due to clean up manual check on UnionTypeMapper, that I found a bug on it.

### CODE REVIEW

File: packages/PHPStanStaticTypeMapper/TypeMapper/UnionTypeMapper.php [hunk 1]
```diff
@@ -284,6 +284,12 @@ private function matchTypeForUnionedTypes(UnionType $unionType, string $typeKind
     {
         $phpParserUnionType = $this->matchPhpParserUnionType($unionType, $typeKind);
 
+        // use first unioned type in case of unioned object types
+        $compatibleObjectTypeNode = $this->processResolveCompatibleObjectCandidates($unionType);
+        if ($compatibleObjectTypeNode instanceof NullableType || $compatibleObjectTypeNode instanceof FullyQualified) {
+            return $compatibleObjectTypeNode;
+        }
```
Comment (samsonasik) [inline packages/PHPStanStaticTypeMapper/TypeMapper/UnionTypeMapper.php:287]: moving compatible object candidate early seems resolve it 👍

**Comment by samsonasik** [comment]
It seems cause a bug on `AddMethodCallBasedStrictParamTypeRector`

```diff
There were 2 failures:

1) Rector\Tests\TypeDeclaration\Rector\ClassMethod\AddMethodCallBasedStrictParamTypeRector\UnionTest::test with data set #1
Failed on fixture file "narrow_union.php.inc"
Failed asserting that string matches format description.
--- Expected
+++ Actual
@@ @@
         $this->someExpr($string);
     }
 
-    private function someExpr(\PhpParser\Node\Expr\MethodCall|\PhpParser\Node\Expr\StaticCall|\PhpParser\Node\Scalar\String_ $expr)
+    private function someExpr(\PhpParser\Node\Expr $expr)
```

I will check more.

**Comment by samsonasik** [comment]
Fixed 🎉

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba I think it is ready.

**Comment by TomasVotruba** [comment]
Looks good to me, thanks :+1:

**Pull request merged**